glotaran/utils/tee.py

```diff
@@ -0,0 +1,80 @@
+"""Module containing ``tee`` like functionality."""
+from __future__ import annotations
+
+import sys
+from io import StringIO
+from types import TracebackType
+
+from glotaran.utils.regex import RegexPattern
+
+
+class TeeContext:
+    """Context manager that allows to work with string written to stdout."""
+
+    def __init__(self) -> None:
+        """Create new ``StringIO`` buffer and save reference to original ``sys.stdout``."""
+        self.buffer = StringIO()
+        self.stdout = sys.stdout
+
+    def __enter__(self) -> TeeContext:
+        """Replace ``sys.stdout`` on entering the context.
+
+        Returns
+        -------
+        TeeContext
+            Instance that can be read from.
+        """
+        sys.stdout = self  # type:ignore[assignment]
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> bool | None:
+        """Restore ``sys.stdout`` on exiting the context."""
+        sys.stdout = self.stdout
+        return None
+
+    def write(self, data: str) -> None:
+        """Write to buffer and original ``sys.stdout``.
+
+        Parameters
+        ----------
+        data: str
+            String to write to stdout and buffer.
+        """
+        self.buffer.write(data)
+        self.stdout.write(data)
+
+    def read(self) -> str:
+        """Return values written to buffer.
+
+        Returns
+        -------
+        str
+            Text written to buffer.
+        """
+        return self.buffer.getvalue()
+
+    def flush(self) -> None:
+        """Flush values in the buffer."""
+        self.buffer.flush()
+
+
+def get_current_optimization_iteration(optimize_stdout: str) -> int:
```

Comment:
This is specific to optimization implementation. Therefore it belongs to the optimizer class. It should be a class method of the Optimizer.

Reply:
Done in [fadb8e4](https://github.com/glotaran/pyglotaran/pull/1134/commits/fadb8e43ab49d1614470794eafb8264f9b079353)